fmt/mod.c

```diff
@@ -396,3 +467,192 @@ int fmt_mod15_load_song(song_t *song, slurp_t *fp, unsigned int lflags)
+				mod_pattern[joutpos + 2] = (m->instrument & 0xf) << 4;
+				mod_fx = 0; mod_fx_val = m->param;
+				if(m->voleffect == VOLFX_VOLUME) {
+					mod_fx = 0x0c;
+					mod_fx_val = m->volparam;
+				} else if (m->voleffect == VOLFX_NONE) {
+					switch(m->effect) {
+						case 0: mod_fx_val = 0; break;
+						case FX_ARPEGGIO: mod_fx = 0; break;
+						case FX_PORTAMENTOUP:
+							mod_fx = 1;
+							if ((mod_fx_val & 0xf0) == 0xe0) {
+								mod_fx = 0x0e;
+								mod_fx_val = 0x10 | ((mod_fx_val & 0xf) >> 2);
+							} else if ((mod_fx_val & 0xf0) == 0xf0) {
+								mod_fx = 0x0e;
+								mod_fx_val = 0x10 | (mod_fx_val & 0xf);
+							}
+							break;
+						case FX_PORTAMENTODOWN:
+							mod_fx = 2;
+							if ((mod_fx_val & 0xf0) == 0xe0) {
+								mod_fx = 0x0e;
+								mod_fx_val = 0x20 | ((mod_fx_val & 0xf) >> 2);
+							} else if ((mod_fx_val & 0xf0) == 0xf0) {
+								mod_fx = 0x0e;
+								mod_fx_val = 0x20 | (mod_fx_val & 0xf);
+							}
+							break;
+						case FX_TONEPORTAMENTO: mod_fx = 3; break;
+						case FX_VIBRATO: mod_fx = 4; break;
+						case FX_TONEPORTAVOL: mod_fx = 5; break;
+						case FX_VIBRATOVOL: mod_fx = 6; break;
+						case FX_TREMOLO: mod_fx = 7; break;
+						case FX_PANNING: mod_fx = 8; break;
+						case FX_OFFSET: mod_fx = 9; break;
+						case FX_VOLUMESLIDE:
+							mod_fx = 0x0a;
+							if ((mod_fx_val & 0xf0) == 0xf0) { // fine volslide down!
+								mod_fx = 0x0e;
+								mod_fx_val &= 0xbf;
+							} else if ((mod_fx_val & 0x0f) == 0x0f) { // fine volslide up!
+								mod_fx = 0x0e;
+								mod_fx_val = 0xa0 | (mod_fx_val >> 4);
+							}
+							break;
+						case FX_POSITIONJUMP: mod_fx = 0x0b; break;
+						case FX_VOLUME: mod_fx = 0x0c; break;
+						case FX_PATTERNBREAK: mod_fx = 0x0d; mod_fx_val = ((mod_fx_val / 10) << 4) | (mod_fx_val % 10); break;
+						case FX_SPEED: mod_fx = 0x0f; break;
+						case FX_TEMPO: mod_fx = 0x0f; break;
+						case FX_SPECIAL:
+							mod_fx = 0x0e;
+							switch (mod_fx_val & 0xf0) {
+								case 0x10: mod_fx_val = (mod_fx_val & 0x0f) | 0x30; break;
+								case 0x20: mod_fx_val = (mod_fx_val & 0x0f) | 0x50; break; // there is an error in Protracker 2.1 docs!
+								case 0x30: mod_fx_val = (mod_fx_val & 0x0f) | 0x40; break;
+								case 0x40: mod_fx_val = (mod_fx_val & 0x0f) | 0x70; break;
+								case 0xb0: mod_fx_val = (mod_fx_val & 0x0f) | 0x60; break;
+								default: break; // handling silently E0x,E6x,E8x,ECx,EDx,EEx,(?EFx)
+							}
+							break;
+						case FX_RETRIG: mod_fx = 0x0e; mod_fx_val = 0x90 | (mod_fx_val & 0x0f); break;
+						default:
+							warn |= 1 << WARN_VOLEFFECTS;
+							break;
+					}
+				} else {
+					warn |= 1 << WARN_VOLEFFECTS;
+				}
+				mod_pattern[joutpos + 2] |= mod_fx & 0x0f;
+				mod_pattern[joutpos + 3] = mod_fx_val;
+				joutpos += 4;
+			}
+		}
+		disko_write(fp, mod_pattern, 1024);
+	}
+
+	// TODO : store samples, truncate to even length per sample //Signed 8-bit mono sample data.
```

Comment:
This TODO absolutely needs to be addressed before merging, there are already enough broken MODs in the wild, and giving people the easy possibility of writing more of them is not helpful.

The way this could be addressed is adding a new `length` parameter to `csf_write_sample`. Existing call sites of the function could simply pass uint32_max as a length, and the function itself would then take `MIN(length, sample->length)` as the length to write.

Reply:
For non-looped samples you should also make sure that the first two bytes are 0 (you can overwrite them after the `csf_write_sample` call), or they will sound broken in ProTracker and other players that implement one-shot loops (technically every MOD sample is looped, but if the first point of the loop is 0, ProTracker first plays the complete sample and then just the looping section. This is why you will typically find loop start 0, loop length 2 for non-looped samples in MOD).

Reply:
done in change @8f2a1fe

Reply:
You'll still need to zero out the first two bytes of the sample if it's not looped. This doesn't need to be part of the csf_write_sample function, you can for example `fseek` back to the beginning of the sample and overwrite the samples afterwards.

Reply:
What do you mean under "zeroing out"? Insertion or overwriting? How do you define "Zero-level": aling to/make them equal to last nibble/byte value of sample, set to signed zeroes, set to unsigned zeroes, repeat first byte etc... Also consider that preparing samples should belong to the end users as well as assuming some Paula knowledge. But IRL, I strongly agree with you ;-)

Reply:
The first two bytes of the sample data should be set to 0. In order to avoid growing the sample each time the file is re-saved, trackers typically just overwrite the first two bytes instead of inserting extra bytes.

> Also consider that preparing samples should belong to the end users as well as assuming some Paula knowledge

No, it absolutely shouldn't. 99% of users don't know about this and are still perfectly capable of making a ProTracker-compatible MOD otherwise. As SchismTracker doesn't even have an easy-to-use sample editor that could help them fixing this, it's absolutely vital that the software automatically does it for them.  I have attached an example file to demonstrate the problem if the tracker doesn't care of zeroing the samples. Load it into ProTracker 2.3d in (Win)UAE and you will notice a high-pitched beep when the sample is finished playing. Most players don't emulate this feature so even if a user tested their MOD file in 10 different MOD players they would still be unlikely to figure out that something is technically wrong with their file.

[oneshot-loop.zip](https://github.com/schismtracker/schismtracker/files/6359823/oneshot-loop.zip)

Reply:
De facto done+pushed your recommended changes. De jure needs testing ;-)

Reply:
Seems to be okay, rebased to schismtracker:master